src/Controller/Component/PaginatorComponent.php

```diff
@@ -187,9 +187,10 @@ public function paginate($object, array $settings = [])
             $query->applyOptions($options);
         }
 
+        $cleanQuery = clone $query;
         $results = $query->all();
         $numResults = count($results);
-        $count = $numResults ? $query->count() : 0;
+        $count = $numResults ? $cleanQuery->count() : 0;
```

Comment:
The issue is that after all() is called the EagerLoader retains state, as the queryBuilder callbacks are modified/wrapped during the first run.